torch/backends/_nnapi/serializer.py

```diff
@@ -1549,11 +1549,28 @@ def add_adaptive_avg_pool2d(self, node):
         self.add_operation(NNAPI_OperationCode.AVERAGE_POOL_2D, inputs, outputs)
 
     def add_upsample_nearest2d(self, node):
-        assert node.inputsSize() == 3
+        assert node.inputsSize() == 3 or node.inputsSize() == 4
```

Comment:
The changes I made to this file are kind of a hacky workaround. The upsample functions, not just `upsample_nearest2d`, have two different overloads that the parser checks for in the corresponding `THPVariable_*` function in `torch/csrc/autograd/generated/python_nn_functions.cpp`. For `upsample_nearest2d`, this used to be how the parser was defined before this PR:

```cpp
  static PythonArgParser parser({
    "upsample_nearest2d(Tensor input, IntArrayRef? output_size, ArrayRef<double>? scale_factors)",
    "upsample_nearest2d(Tensor input, IntArrayRef[2] output_size, double? scales_h=None, double? scales_w=None, *, Tensor out=None)",
  }, /*traceable=*/true);
```

If `upsample_nearest2d` is called with just the `input` and `output_size`, and no other arguments, both of function schemas are a match, but the parse checks them in order, so the first one would be chosen. That worked fine, until this PR changed `IntArrayRef?` to `OptionalIntArrayRef`. These entries are sorted in alphabetical order, so now they are switched:

```cpp
  static PythonArgParser parser({
    "upsample_nearest2d(Tensor input, IntArrayRef[2] output_size, double? scales_h=None, double? scales_w=None, *, Tensor out=None)",
    "upsample_nearest2d(Tensor input, OptionalIntArrayRef output_size, ArrayRef<double>? scale_factors)",
  }, /*traceable=*/true); 
```

And now the wrong overload is chosen. The one with `scales_h` and `scales_w` is an old overload that still exists for BC, and it doesn't have all the features that the newer overload has.

One of the missing features was support for this NNAPI serializer. The change I made in this file adds support for that, and it seems to be working. It looks like `upsample_nearest2d` was the only upsample function that was supported here.

However, there's another missing feature that won't be so simple (nor advisable, probably) to add for these old overloads--forward AD. At the moment, all of the upsample functions are failing forward AD tests because of the overload reordering. If I could just change the order back to what it used to be, the issue should be fixed, but at the moment, I'm not completely sure how to do that. We could add specialized logic in codegen to reorder the upsample functions, but I'm not sure if that's a good idea.

@peterbell10 or @mruberry, do you have any thoughts?

Reply:
I just tried the codegen reordering idea, and it seems to work. I just reversed the `grouped_overloads` within the `method_impl()` function in `tools/autograd/gen_python_functions.py` any time it sees an upsample function:
https://github.com/pytorch/pytorch/blob/d8abe813bc7ae02d0cbdfdc8fa8db0d17cbe8ae8/tools/autograd/gen_python_functions.py#L592

It looks like there already are some special-cased comparisons in the sorting function here: https://github.com/pytorch/pytorch/blob/d8abe813bc7ae02d0cbdfdc8fa8db0d17cbe8ae8/tools/autograd/gen_python_functions.py#L879

So I guess it might not be a bad idea to just add some logic here to prioritize `OptionalIntArrayRef` over `IntArrayRef`, or something like that.

Reply:
I think all ambiguous overloads should behave equivalently; so I would consider this a bug in the nnapi code. The `scales_h` overload even looks to be the main overload in ATen, the other one just wraps it.

Also, while I just said it shouldn't matter. `IntArrayRef` is a more specific overload than `OptionalIntArrayRef` so it seems backwards to sort the `Optional` version first.

Reply:
That's a fair point. Then I guess we should enable forward AD for the`IntArrayRef` overloads?